File: week_1/yoon/sw_4866.py
T = int(input())
# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    line = input()
    n = len(line)
    stack = []

    for l in range(n):
        if line[l] == "(" or line[l]  == "{":
            stack.append(line[l])
        elif line[l] == ")" or line[l] == "}":
            if len(stack) == 0:
                stack = [line[l]]
                break
            elif (line[l] == "}" and stack[-1] != "{") or (line[l] == ")" and stack[-1] != "("):
                stack = [line[l]]
                break
            else:
                stack.pop()

 	
    if len(stack) == 0:
        print("#{} {}".format(test_case, 1))
    else:
        print("#{} {}".format(test_case, 0))

File: week_1/yoon/sw_4866.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int solution(string line)
{
    stack<char> s;
    for(int j = 0; j < line.size(); ++j)
    {
        if(line[j] == '(' | line[j] == '{')
            s.push(line[j]);
        
        else if(line[j] == ')' | line[j] == '}')
        {
            if(s.size() == 0)
            {
                s.push(line[j]);
                break;
            }
            else if((line[j] == ')' && s.top() != '(') | (line[j] == '}' && s.top() != '{'))
            {
                s.push(line[j]);
                break;
            }
            else
            {
                s.pop();
            }
        }   
    }
    if(s.size() == 0)
        return 1;
    else
        return 0;
    
}


int main()
{
    int t;
    cin >> t;
    int result;
    std::string line;
    std::getline(cin, line);
    for(int i = 1; i <= t; ++i)
    {        
        std::getline(cin, line);
        result = solution(line);

        cout << "#" << i << " " << result << endl;

    }

    //cout << result;
} 